Add support for ListTensor to ExpressionBuilder.

kynan opened the pull request:
A ListTensor wraps a list of expressions into a tensor valued
expression of one higher rank. This commit adds supported for these to
MCFC.

When a ListTensor is encountered in the expression tree, capture its
subordinated expressions and build an initialiser for a fixed size of
the dimension of the ListTensor with these expressions. Push this
initialiser on a subexpression stack.

Subscript this array with an index over dimensions and push that on
the expression stack. In the loop over dimensions for the expression,
the appropriate expression of the ListTensor is thereby accessed.

Add a test case based on the Euler-advection test.

Note: This has only been tested with scalar expressions below the
ListTensor and may well break for higher rank.

kynan committed:
Ignore function definitions (FunctionDef) and imports (ImportFrom) in ASTVisualiser.

kynan committed:
Add support for ListTensor to ExpressionBuilder.

A ListTensor wraps a list of expressions into a tensor valued
expression of one higher rank. This commit adds supported for these to
MCFC.

When a ListTensor is encountered in the expression tree, capture its
subordinated expressions and build an initialiser for a fixed size of
the dimension of the ListTensor with these expressions. Push this
initialiser on a subexpression stack.

Subscript this array with an index over dimensions and push that on
the expression stack. In the loop over dimensions for the expression,
the appropriate expression of the ListTensor is thereby accessed.

Note: This has only been tested with scalar expressions below the
ListTensor and may well break for higher rank.

kynan committed:
Add swapped-advection test case.

gmarkall commented:
This looks good - I really like how simple a solution it is to the problem! I had been dreading looking into it.

The only thing that I think we're missing is swapped-advection.ufl

kynan committed:
Remove the expression stack from the expression builder and have the visitors return the expressions they build instead.

kynan committed:
Generalise ListTensor visitor for the higher rank case.

kynan committed:
Add a division operator.

kynan committed:
Use IndexSum indices to build and subscript ListTensor.

kynan committed:
[minor] codegeneration: change getter function names for Variable to match those of other expressions

kynan committed:
Partitioner uses IndexSum indices rather than only their depth.

Instead of creating a static loop nest for expressions with all
IndexSum indices found, have the Partitioner find the IndexSum indices
per partition. Use those to create an appropriate loop nest for the
partition, reusing previously created loops from other partitions if
they use the same indices.

This doesn't always create the ideal loop nest, since IndexSum indices
have different counts in different branches of the tree despite having
a sub expression of the same form beneath.

The IndexSumCounter is obsolete and removed.
